### DISCUSSION TITLE
How does one use a request response as a global dictionary / mapping AND array?

### DISCUSSION THREAD

**Discussion by iron3oxide** [lowdefy/lowdefy#1270] [Questions and Help] [answered]
This is a bit of a complex question, therefore I'll add some context: 
My app uses a form where users (amongst other stuff) input nations in an AutoComplete field. Next to said field is a non-editable text field where (once a nation has been input), the corresponding abbreviation should appear, e.g. "FR" for France. 

On the technical side, I want to have an API endpoint that gets called once on a sensible event, e.g. when the app loads. This returns a json response following the schema:

```
{...
 "France": {"id": "23", "abbreviation": "FR"},
 "Belgium": {"id": "24", "abbreviation": "BE"},
 ...
 }                 
```

I then want to 
A) use the array of keys/country names of that response to serve as the options for the aforementioned AutoComplete field
and
B) use the response as a global dictionary / mapping to automatically fill in the correct abbreviation and to send the nation id instead of the name back to my API on POST requests.

From #466 and other discussions I gather that the latter can be achieved by using SetState (so the API doesn't get called each time), _nunjucks (so the nation name can be a variable) and referencing the keys in the data object by name, e.g. nations.{{ nation_name }}.id 

UPDATE:
I just found the _object.keys method, I guess that's what I should use for A)...

Is this feasible? If so, how? Or is it feasible but my design is bad? Thanks in advance and keep up the great work!

**Comment by StephanieJKS** [comment, marked as answer]
Hi @iron3oxide,
Yes it totally is!
There are multiple ways to achieve this, here is one option:

```yaml
(page that calls the api to get the data for request id "country_codes")
  - id: country
    type: Selector
    # NOTE:
    # A Selector would be a better choice here, if the user should only be
    # able to select a country that appears in the json response.
    # AutoComplete allows the user to enter an option that is not part
    # of the list and therefore one that is not in the json response.
    # https://docs.lowdefy.com/Selector
    layout:
      span: 16
    properties:
      options:
        _array.map:
          on:
            _object.keys:
              _request: country_codes
          # NOTE: In the data format.
          # {...
          # "France": {"id": "23", "abbreviation": "FR"},
          # "Belgium": {"id": "24", "abbreviation": "BE"},
          # ...
          # }
          #  Check if this is this is your format with
          #   _log:
          #     _request: country_codes
          # which will log the request response to the browser console.
          # If you are using AxiosHttp you will likely need to destruct,
          # for example _request: country_codes.data
          callback:
            _function:
                label:
                  args: 0
                value:
                  get:
                    from:
                      _request: country_codes
                    key:
                      __args: 0
    events:
      # NOTE:
      # Here we make use of SetState to update the state of the TextInput block "code"
      # to the corresponding abbreviation of the selected country.
      # The corresponding id can be retrieved using _state: country.id
      onChange:
        - id: update_code
          type: SetState
          params:
            code:
              _state: country.abbreviation 
  - id: code
    type: TextInput
    layout:
      span: 8
    properties:
      disabled: true

**Reply by iron3oxide** [reply]
Thanks so much, very detailed and helpful!
